Implemented TestExecutionContext to use AsyncLocal<> for NETSTANDARD1_6

Suremaker opened the pull request:
This is a partial fix for #1965.
I tested the change later with project using `net46` as well as `netcoreapp1.0`.
The pull request contains also test covering #1952

CharliePoole commented:
Would this be a partial fix to #1965?

Suremaker commented:
What I understand the `NetStandard` change would cover `netcoreapp1.0` and other frameworks which are not generic `.NETFramework`.

I checked in added tests that implementation which is already present for everything else than `PORTABLE` and `NETSTANDARD1_6` was actually working fine.

So I believe that the only problematic scenario now would be `PORTABLE` implementation.

Suremaker commented:
I have managed to make my tests passing for both `net46` and `netcoreapp1.0`.

jnm2 commented:
According to [apisof.net](https://apisof.net/catalog/System.Threading.AsyncLocal%3CT%3E), `AsyncLocal` is available to `netcore10`, `netstandard13`, `net46`, `win10`, and newer.

CharliePoole commented:
@Suremaker So I guess that's a yes. :smile: 

We need to figure out an additional solution that covers .NET 2.0 through 4.5. But I'll review this as a fix for at least certain platforms.

oznetmaster commented:
Would using Thread.GetData/SetData work on all versions of .NET correctly for async?

Suremaker commented:
@CharliePoole I think that `CallContext.SetData(CONTEXT_KEY, context); / CallContext.GetData(CONTEXT_KEY);` solution is enough for .net 4.5 - the tests I have added covers net4.5 and they are passing.
I did not add any tests for .net 2.0 - 4.0 as they do not have async/await constructs, but it should be possible to write tests using tasks to verify if starting from net4.0.
The potentially biggest issue would be to implement it properly in Portable version - currently I have no idea how to do it better for it.
Also, it would be probably worth to add NET46 version and ensure that it uses AsyncLocal<> ?

@oznetmaster As far I am aware `Thread.GetData/SetData` uses thread local storage so it will not work with async/await in my opinion.

oznetmaster commented:
According to this article, that will not work (CallContext.SetData/GetData): http://stackoverflow.com/questions/13010563/using-threadstatic-variables-with-async-await

jnm2 commented:
> Also, it would be probably worth to add NET46 version and ensure that it uses `AsyncLocal<>` ?

I think so, either that or build for .NET Standard 1.3 and run tests on the .NET Framework runtime 4.6.
This whole multi-targeting thing is about to get ridiculously easier with MSBuild 15.

jnm2 commented:
@oznetmaster is right, but it should be possible to write a test proving this.

The weird part is we have async language support polyfillsbback to .NET Framework 3.5 but don't they come with an AsyncLocal?

Suremaker commented:
@oznetmaster @jnm2 It really look like CallContext.SetData/GetData should not work but so far all the tests I created passed so far. There is an other article (linked also on stackoverflow) describing that CallContext.LogicalSetData/GetData could make the work for .net 45: http://blog.stephencleary.com/2013/04/implicit-async-context-asynclocal.html
So I imagine it would be good to write some tests recreating this issue with CallContext.SetData/GetData and try to fix it with other approach.

rprouse commented:
>I did not add any tests for .net 2.0 - 4.0 as they do not have async/await constructs, but it should be possible to write tests using tasks to verify if starting from net4.0.

Our 4.0 framework and its tests reference the BCL library and have async/await support, so you should be able to enable your tests for the 4.0 framework.

Suremaker commented:
I have updated tests to cover .net40 and added two more complicated scenarios where async method branches and executes it's bits in parallel.
All the tests came green on my box - also because of this reason I have not changed `CallContext.SetData/GetData` as I cannot prove them not working.

The last scenario that comes to my mind is to run those tests on a machine with only .net40 installed - maybe it will reveal issues with CallContext that I cannot reproduce as my machine has .net46 which is in-place replacement for .net45 / .net40 (and it may changed the behaviour of those classes). I do not have however any box with .net40 only to try it out.

Suremaker commented:
Is there anything else I should do here in order to accept the pull request?

rprouse commented:
@Suremaker we are in the middle of a release cycle, so it will take us a bit longer than normal to review pull requests and move things along. There is a lot going on in the various projects right now. Please bare with us 😄

CharliePoole commented on src/NUnitFramework/tests/TestContextTests.cs:
```diff
@@ -233,6 +237,123 @@ public void TestContextStoresFailureInfoForOneTimeTearDown()
+#if ASYNC
+
+#if PORTABLE
+        [Ignore("Not implemented yet as TestExecutionContext uses TLS in portable implementation")]
```
My first thought was that you should use a PlatformAttribute here, but portable isn't really a platform but a combination of platforms. Not sure how to deal with that any better than you have done. 😞

rprouse commented:
@CharliePoole looks like we were reviewing at the same time. This looks good to me. I am pretty sure we will drop PCL in favor of NET Standard 1.0 or 1.1, so that omission doesn't worry me. Does the @nunit/core-team agree with that?
